### PR TITLE
feat: support IPFS/IPNS paths, Gateways, etc

### PR DESCRIPTION
**Pull request by achingbrain**
Adds support for fetching resources in the form:

- IPFS Path, e.g. `/ipfs/<CID>`
- IPNS Path, e.g. `/ipns/<PeerId>`
- IPFS Gateway Path, e.g. `http://example.com/ipfs/<CID>`
- IPNS Gateway Path, e.g. `http://example.com/ipns/<PeerId>`
- IPFS Subdomain Gateway Path, e.g. `http://<CID>.ipfs.example.com`
- IPNS Subdomain Gateway Path, e.g. `http://<PeerId>.ipns.example.com`

Notes:

- Supports paths & queries as per `ipfs://` and `ipns://` URLs.
- Uses all available methods to resolve a given CID to some content - it does not, for example, when given a subdomain gateway path, restrict downloading the blocks from only that subdomain gateway

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation if necessary (this includes comments as well)
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by achingbrain** [comment]
After https://github.com/ipfs/helia-verified-fetch/pull/5 we will have the `.url` property on responses that is the canonical URL for the requested resource.

With this PR the user can request the same resource via multiple URL formats.  Should we be rewriting this URL to turn, for example, subdomain gateway requests into `ipfs://` URIs?

File: packages/verified-fetch/src/utils/parse-url-string.ts [hunk 1]
```diff
@@ -31,6 +31,21 @@ export interface ParsedUrlStringResults {
 }
 
 const URL_REGEX = /^(?<protocol>ip[fn]s):\/\/(?<cidOrPeerIdOrDnsLink>[^/$?]+)\/?(?<path>[^$?]*)\??(?<queryString>.*)$/
+const PATH_REGEX = /^\/(?<protocol>ip[fn]s)\/(?<cidOrPeerIdOrDnsLink>[^/$?]+)\/?(?<path>[^$?]*)\??(?<queryString>.*)$/
```
Comment (2color) [inline packages/verified-fetch/src/utils/parse-url-string.ts:34]: Why are the `$` necessary in the negated sets of the named capture groups, i.e., `[^/$?]` are we expecting $ in the URL?

File: packages/verified-fetch/src/utils/parse-url-string.ts [hunk 2]
```diff
@@ -31,6 +31,21 @@ export interface ParsedUrlStringResults {
...
+const PATH_GATEWAY_REGEX = /^http(s?):\/\/(.*)\/(?<protocol>ip[fn]s)\/(?<cidOrPeerIdOrDnsLink>[^/$?]+)\/?(?<path>[^$?]*)\??(?<queryString>.*)$/
```
Comment (2color) [inline packages/verified-fetch/src/utils/parse-url-string.ts:35]:
```suggestion
const PATH_GATEWAY_REGEX = /^http[s?]:\/\/(.*[^\/])\/(?<protocol>ip[fn]s)\/(?<cidOrPeerIdOrDnsLink>[^/$?]+)\/?(?<path>[^$?]*)\??(?<queryString>.*)$/
```

You probably don't need to capture the s in http and avoid matching everything in the hostname part of the URL.

File: packages/verified-fetch/src/utils/parse-url-string.ts [hunk 3]
```diff
@@ -31,6 +31,21 @@ export interface ParsedUrlStringResults {
...
+const SUBDOMAIN_GATEWAY_REGEX = /^http(s?):\/\/(?<cidOrPeerIdOrDnsLink>[^/$?]+)\.(?<protocol>ip[fn]s)\.([^/?]*)\/?(?<path>[^$?]*)\??(?<queryString>.*)$/
```
Comment (2color) [inline packages/verified-fetch/src/utils/parse-url-string.ts:36]:
```suggestion
const SUBDOMAIN_GATEWAY_REGEX = /^http(s?):\/\/(?<cidOrPeerIdOrDnsLink>[^/$\.?]+)\.(?<protocol>ip[fn]s)\.([^/?]*)\/?(?<path>[^$?]*)\??(?<queryString>.*)$/
```

File: packages/verified-fetch/src/utils/parse-url-string.ts [hunk 4]
```diff
@@ -31,6 +31,21 @@ export interface ParsedUrlStringResults {
...
+function matchURLString (urlString: string): Record<string, string> {
+  for (const pattern of [URL_REGEX, PATH_REGEX, PATH_GATEWAY_REGEX, SUBDOMAIN_GATEWAY_REGEX]) {
+    const match = urlString.match(pattern)
+
+    if (match?.groups != null) {
+      return match.groups
+    }
+  }
+
+  throw new TypeError(`Invalid URL: ${urlString}, please use ipfs:// or ipns:// URLs only.`)
```
Comment (2color) [inline packages/verified-fetch/src/utils/parse-url-string.ts:47]:
```suggestion
  throw new TypeError(`Invalid URL: ${urlString}, please use ipfs://, ipns://, or gateway URLs only.`)
```

File: packages/verified-fetch/src/utils/parse-url-string.ts (hunk 2 above)
Comment (achingbrain) [reply packages/verified-fetch/src/utils/parse-url-string.ts:35]: I don't think the `[s?]` is necessary either, just `?` seems to work.

File: packages/verified-fetch/src/utils/parse-url-string.ts (hunk 1 above)
Comment (achingbrain) [reply packages/verified-fetch/src/utils/parse-url-string.ts]: I'm not sure why they are there, it doesn't look like they are necessary and the tests pass without them, have removed.

File: packages/verified-fetch/src/utils/parse-url-string.ts (hunk 3 above)
Comment (achingbrain) [reply packages/verified-fetch/src/utils/parse-url-string.ts]: I've made this change, though without the `$`

**Comment by SgtPooki** [comment]
> With this PR the user can request the same resource via multiple URL formats. Should we be rewriting this URL to turn, for example, subdomain gateway requests into ipfs:// URIs?

I think we would want to just parse the information we need.. 

---

some comments unrelated to the above, but more for the change this PR brings in general:

I was hoping this library could stick with `ipfs://` and `ipns://` as strictly as possible so we could more officially advocate for these protocol urls until we got native browser support, but the DX argument for supporting https & http urls is pretty strong. This change will make using this library much simpler for consumers.

File: packages/verified-fetch/src/utils/parse-url-string.ts [hunk 5]
```diff
@@ -41,13 +56,7 @@ const URL_REGEX = /^(?<protocol>ip[fn]s):\/\/(?<cidOrPeerIdOrDnsLink>[^/$?]+)\/?
...
 export async function parseUrlString ({ urlString, ipns, logger }: ParseUrlStringInput, options?: ParseUrlStringOptions): Promise<ParsedUrlStringResults> {
   const log = logger.forComponent('helia:verified-fetch:parse-url-string')
-  const match = urlString.match(URL_REGEX)
-
-  if (match == null || match.groups == null) {
-    throw new TypeError(`Invalid URL: ${urlString}, please use ipfs:// or ipns:// URLs only.`)
-  }
-
-  const { protocol, cidOrPeerIdOrDnsLink, path: urlPath, queryString } = match.groups
+  const { protocol, cidOrPeerIdOrDnsLink, path: urlPath, queryString } = matchURLString(urlString)
```
Comment (SgtPooki) [inline packages/verified-fetch/src/utils/parse-url-string.ts:59]: nice improvement here

File: packages/verified-fetch/test/utils/parse-url-string.spec.ts [hunk 6]
```diff
@@ -357,15 +566,275 @@ describe('parseUrlString', () => {
...
+            query: {
+              format: 'tar'
+            }
+          }
+        )
+      })
+    })
+  })
+
+  HTTP_PROTOCOLS.forEach(proto => {
```
Comment (SgtPooki) [inline packages/verified-fetch/test/utils/parse-url-string.spec.ts:754]: I was `CTRL+F`ing for `http` and `https` and it took me way too long to find this.

File: packages/verified-fetch/test/utils/parse-url-string.spec.ts [hunk 7]
```diff
@@ -357,15 +566,275 @@ describe('parseUrlString', () => {
...
+    describe(`${proto}://<CID>.ipns.example.com URLs`, () => {
```
Comment (SgtPooki) [inline packages/verified-fetch/test/utils/parse-url-string.spec.ts:755]:
```suggestion
    describe(`${proto}://<key>.ipns.example.com URLs`, () => {
```

**Review by SgtPooki — APPROVED** [review]
minor nit on test group name, I like that it didn't take much code (besides a ton of tests) to add this support though!

**Pull request merged**